Switch to TF 2.0 and new NLU components (#5266)
RasaHQ/rasa: docs/core/policies.rst

Embedding Policy
^^^^^^^^^^^^^^^^

    .. warning::

        ``EmbeddingPolicy`` got renamed to ``TEDPolicy``. Please use :ref:`ted_policy` instead of ``EmbeddingPolicy``.
        The functionality of the policy stayed the same.

.. _ted_policy:

TED Policy
^^^^^^^^^^

Transformer Embedding Dialogue (TED) Policy is described in
`our paper <https://arxiv.org/abs/1910.00486>`__.

This policy has a pre-defined architecture, which comprises the
following steps:

    - concatenate user input (user intent and entities), previous system actions, slots and active forms for each time
      step into an input vector to pre-transformer embedding layer;
    - feed it to transformer;
    - apply a dense layer to the output of the transformer to get embeddings of a dialogue for each time step;
    - apply a dense layer to create embeddings for system actions for each time step;
    - calculate the similarity between the dialogue embedding and embedded system actions.
      This step is based on the `StarSpace <https://arxiv.org/abs/1709.03856>`_ idea.

> It is recommended to use ``state_featurizer=LabelTokenizerSingleStateFeaturizer(...)``

Reviewer: this is from the old docs

> It is recommended to use ``state_featurizer=LabelTokenizerSingleStateFeaturizer(...)``

Reviewer 2: yes but it came up here that it doesn't make sense

> It is recommended to use ``state_featurizer=LabelTokenizerSingleStateFeaturizer(...)``

Reviewer: did make sense before?

> It is recommended to use ``state_featurizer=LabelTokenizerSingleStateFeaturizer(...)``

Reviewer 2: no

> It is recommended to use ``state_featurizer=LabelTokenizerSingleStateFeaturizer(...)``

Author: so what should we do?

Embedding Policy
^^^^^^^^^^^^^^^^

    .. warning::

        ``EmbeddingPolicy`` got renamed to ``TEDPolicy``. Please use :ref:`ted_policy` instead of ``EmbeddingPolicy``.
        The functionality of the policy stayed the same.

> 

Reviewer 3: ```suggestion
```

.. _ted_policy:

TED Policy
^^^^^^^^^^

> Transformer Embedding Dialogue Policy (TEDP)

Reviewer 4: don't we just call it TED rather than TEDP?

Author: changed it to `Transformer Embedding Dialogue (TED) Policy`

> Transformer Embedding Dialogue (TED) Policy

Reviewer 3: it's not a 100% clear to mee what these words / sentence is for?

> The policy used in our paper https://arxiv.org/abs/1910.00486.

Reviewer 3: Please make it a real link

This policy has a pre-defined architecture, which comprises the
following steps:

>     - concatenate user input (user intent and entities), previous system action, slots and active form for each time

Reviewer 3: ```suggestion
    - concatenate user input (user intent and entities), previous system actions, slots and active forms for each time
```

>       step into an input vector to pre-transformer embedding layer;

Reviewer 3: can we keep these explanations a bit more high level? If somebody wants to know the details they should refer to the paper in my opinion.

Author: will be part of a separate PR https://github.com/RasaHQ/rasa/issues/5308

    - feed it to transformer;
    - apply a dense layer to the output of the transformer to get embeddings of a dialogue for each time step;
    - apply a dense layer to create embeddings for system actions for each time step;
    - calculate the similarity between the dialogue embedding and embedded system actions.
      This step is based on the `StarSpace <https://arxiv.org/abs/1709.03856>`_ idea.

It is recommended to use ``state_featurizer=LabelTokenizerSingleStateFeaturizer(...)``
(see :ref:`featurization_conversations` for details).

**Configuration:**

    Configuration parameters can be passed as parameters to the
    ``TEDPolicy`` within the configuration file.

    .. warning::

>         Pass an appropriate number of ``epochs`` to the ``TEDPolicy``, otherwise the policy will be trained only

Reviewer 3: what is an "appropriate number"?

Author: will be part of a separate PR https://github.com/RasaHQ/rasa/issues/5308

        for ``1`` epoch.

    The algorithm also has hyper-parameters to control:

        - neural network's architecture:

            - ``hidden_layers_sizes`` sets a list of hidden layers
              sizes before embedding layer for system actions, the number
              of hidden layers is equal to the length of the list.
            - ``transformer_size`` sets the number of units in the transfomer.

>             - ``number_of_transformer_layers`` sets the number of transformer layers.

Reviewer 3: I agree. Users who want to improve their bot by trying out different options will not necessarily be data scientists / researchers. We should point out what the most important settings are (or what we think the most important ones are) and explain the impact of each setting.

>             - ``number_of_transformer_layers`` sets the number of transformer layers.

Author: What do you expect from these docs? To teach non data scientists / researchers to improve their docs we need to add much more explanations. We think that should go into a separate PR and should not be part of this PR as it goes into the direction of educating our community. Also there are no options that magical improve the performance of your bot whenever you change them. 
@akelad What do you think?

>             - ``number_of_transformer_layers`` sets the number of transformer layers.

Author: https://github.com/RasaHQ/rasa/issues/5308

            - ``maximum_sequence_length`` sets maximum sequence length.
            - ``number_of_attention_heads`` sets the number of heads in multihead attention.

>             - ``use_key_relative_attention`` if true use key relative embeddings in attention.

Reviewer 3: ```suggestion
            - ``use_key_relative_attention`` if `True` use key relative embeddings in attention.
```

        - regularization:

            - ``regularization_constant`` sets the scale of L2 regularization.
            - ``negative_margin_scale`` sets the scale of how important is to minimize
              the maximum similarity between embeddings of different
              intent labels, used only if ``loss_type`` is set to ``margin``.

>             - ``droprate_dialogue`` sets the dropout rate between

Reviewer 3: It's `drop rate`, isn't it? Same for the ones below
```suggestion
            - ``drop_rate_dialogue`` sets the dropout rate between
```

        # If 'True' the algorithm only minimizes maximum similarity over
        # incorrect intent labels, used only if 'loss_type' is set to 'margin'.
        "use_maximum_negative_similarity": True
        # Scale loss inverse proportionally to confidence of correct prediction
        "scale_loss": True
        # ## Regularization parameters
        # The scale of regularization
        "regularization_constant": 0.001
        # The scale of how important is to minimize the maximum similarity
        # between embeddings of different labels.
        "negative_margin_scale": 0.8
        # Dropout rate for embedding layers of dialogue features.
        "drop_rate_dialogue": 0.1
        # Dropout rate for embedding layers of label, e.g. action, features.
        "drop_rate_label": 0.0
        # Dropout rate for attention.
        "drop_rate_attention": 0
        # ## Evaluation parameters
        # How often calculate validation accuracy.
        # Small values may hurt performance, e.g. model accuracy.
        "evaluate_every_number_of_epochs": 20
        # How many examples to use for hold out validation set
        # Large values may hurt performance, e.g. model accuracy.

>         "evaluate_on_number_of_examples": 0

Author: would keep it here

        ``batch_size`` is required, pass an ``int``, e.g.
        ``"batch_size": 8``.

    These parameters can be specified in the configuration file.
    The following default values are set:

    .. code-block:: yaml

>         # nn architecture

Reviewer 3: ```suggestion
        # Architecture of the used neural network.
```

>         # a list of hidden layers sizes before dialogue and action embed layers

Reviewer 3: ?
```suggestion
        # a list of hidden layers sizes before dialogue and action embedding layers
```

Reviewer 3: if this is copied from the code then we should do this code include tags instead of duplicating it.

Author: We are using constants in code. They are not replaced when included here via the code include tags. As the constant names might be different from the values, we add the defaults manually. Let me know if you know a better way on how to do this.

        # the number of incorrect labels, the algorithm will minimize
        # their similarity to the user input during training
        "use_maximum_negative_similarity": True  # flag which loss function to use
        # scale loss inverse proportionally to confidence of correct prediction
        "scale_loss": True
        # regularization
        # the scale of regularization
        "regularization_constant": 0.001
        # the scale of how important is to minimize the maximum similarity
        # between embeddings of different labels
        "negative_margin_scale": 0.8
        # dropout rate for dial nn
        "droprate_dialogue": 0.1
        # dropout rate for bot nn
        "droprate_label": 0.0
        # dropout rate for attention
        "droprate_attention": 0
        # visualization of accuracy
        # how often calculate validation accuracy
        "evaluate_every_number_of_epochs": 20  # small values may hurt performance
        # how many examples to use for hold out validation set
        "evaluate_on_number_of_examples": 0  # large values may hurt performance

    .. note::

>           Parameter ``maximum_negative_similarity`` is set to a negative value to mimic

Reviewer 3: ```suggestion
          The parameter ``maximum_negative_similarity`` is set to a negative value to mimic
```
